File: android/app/src/main/java/org/urlguardian/app/core/Confusables.kt
package org.urlguardian.app.core

import java.net.IDN
import java.text.Normalizer
import java.util.Locale

object Confusables {
    private val map: Map<Char, Char> = mapOf(
        'а' to 'a', 'в' to 'b', 'е' to 'e', 'ё' to 'e', 'к' to 'k', 'м' to 'm', 'н' to 'h', 'о' to 'o',
        'р' to 'p', 'с' to 'c', 'т' to 't', 'у' to 'y', 'х' to 'x', 'ѕ' to 's', 'і' to 'i', 'ј' to 'j',
        'ԁ' to 'd', 'ɡ' to 'g', 'ь' to 'b', 'г' to 'r', 'п' to 'n',
        'α' to 'a', 'β' to 'b', 'ε' to 'e', 'ι' to 'i', 'κ' to 'k', 'ο' to 'o', 'ρ' to 'p', 'τ' to 't',
        'υ' to 'u', 'ν' to 'v', 'μ' to 'u', 'χ' to 'x', 'γ' to 'y',
        'ł' to 'l', 'ø' to 'o', 'đ' to 'd', 'ı' to 'i', 'ſ' to 's',
        '0' to 'o', '1' to 'l', '3' to 'e', '4' to 'a', '5' to 's', '6' to 'b', '7' to 't', '8' to 'b', '9' to 'g',
    )
    private val digraphs = listOf("rn" to "m", "vv" to "w")
    private val splitter = Regex("[^\\p{L}\\p{N}]+")

    const val MIN_ALIAS_LENGTH = 4

    fun skeleton(value: String): String {
        val decomposed = Normalizer.normalize(value, Normalizer.Form.NFKD).lowercase(Locale.ROOT)
        val output = StringBuilder()
        for (character in decomposed) {
            map[character]?.let { output.append(it); continue }
            val type = Character.getType(character)
            if (type == Character.NON_SPACING_MARK.toInt() ||
                type == Character.COMBINING_SPACING_MARK.toInt() ||
                type == Character.ENCLOSING_MARK.toInt()
            ) continue
            if (Character.isLetterOrDigit(character)) output.append(character)
        }
        return output.toString()
    }

    fun skeletonVariants(value: String): Set<String> {
        val variants = mutableSetOf(skeleton(value))
        for ((source, target) in digraphs) {
            val additions = variants.map { it.replace(source, target) }
            variants.addAll(additions)
        }
        return variants
    }

    fun labelTokens(value: String): Set<String> {
        val tokens = mutableSetOf<String>()
        for (part in value.replace('_', ' ').split(splitter)) {
            for (variant in skeletonVariants(part)) {
                if (variant.length >= MIN_ALIAS_LENGTH) tokens.add(variant)
            }
        }
        return tokens
    }

    fun looksConfusable(value: String): Boolean = value.lowercase(Locale.ROOT).any { !it.isAscii() || it.isDigit() }

    fun idnaDecode(label: String): String {
        if (!label.startsWith("xn--")) return label
        return try { IDN.toUnicode(label) } catch (_: Exception) { label }
    }

    private fun Char.isAscii(): Boolean = code < 128
}

File: src/intel/confusables.py
"""Deterministic confusable-skeleton normalisation for brand matching.

This is a documented heuristic for homoglyph and leetspeak obfuscation
(Cyrillic/Greek lookalikes, digit substitutions such as `0`->`o` and `1`->`l`,
and the `rn`->`m` / `vv`->`w` digraphs). It is used only to *detect* a brand
token in a hostname, never to allow or block by itself, and it is not a
security guarantee: a domain that survives the skeleton is not proven benign.
"""

from __future__ import annotations

import re
import unicodedata

CONFUSABLE_MAP: dict[str, str] = {
    # Cyrillic lookalikes
    "а": "a", "в": "b", "е": "e", "ё": "e", "к": "k", "м": "m", "н": "h", "о": "o",
    "р": "p", "с": "c", "т": "t", "у": "y", "х": "x", "ѕ": "s", "і": "i", "ј": "j",
    "ԁ": "d", "ɡ": "g", "ь": "b", "г": "r", "п": "n",
    # Greek lookalikes
    "α": "a", "β": "b", "ε": "e", "ι": "i", "κ": "k", "ο": "o", "ρ": "p", "τ": "t",
    "υ": "u", "ν": "v", "μ": "u", "χ": "x", "γ": "y",
    # Latin lookalikes and leetspeak digits
    "ł": "l", "ø": "o", "đ": "d", "ı": "i", "ſ": "s",
    "0": "o", "1": "l", "3": "e", "4": "a", "5": "s", "6": "b", "7": "t", "8": "b", "9": "g",
}

DIGRAPH_MAP: tuple[tuple[str, str], ...] = (("rn", "m"), ("vv", "w"))

MIN_ALIAS_LENGTH = 4


def skeleton(value: str) -> str:
    """Lowercase alphanumeric skeleton with confusables mapped to ASCII."""

    decomposed = unicodedata.normalize("NFKD", value)
    output: list[str] = []
    for character in decomposed.lower():
        if character in CONFUSABLE_MAP:
            output.append(CONFUSABLE_MAP[character])
            continue
        if unicodedata.combining(character):
            continue
        if character.isalnum():
            output.append(character)
    return "".join(output)


def skeleton_variants(value: str) -> set[str]:
    """Skeleton plus digraph-collapsed variants used for comparison only."""

    base = skeleton(value)
    variants = {base}
    for source, target in DIGRAPH_MAP:
        additions = {item.replace(source, target) for item in variants}
        variants |= additions
    return variants


def label_tokens(value: str) -> set[str]:
    """Token set for one hostname label or free-text fragment.

    Tokens are split on any non-alphanumeric character, so `paypal-secure`
    yields `{"paypal", "secure"}` and `applepie` yields `{"applepie"}`.
    """

    tokens: set[str] = set()
    for part in re.split(r"\W+", value.replace("_", " ")):
        for variant in skeleton_variants(part):
            if len(variant) >= MIN_ALIAS_LENGTH:
                tokens.add(variant)
    return tokens


def looks_confusable(value: str) -> bool:
    """True when the value contains non-ASCII or digit characters that the
    skeleton had to rewrite, i.e. the match depended on confusable handling."""

    for character in value.lower():
        if not character.isascii() or character.isdigit():
            return True
    return False


def idna_decode(label: str) -> str:
    """Best-effort punycode label decoding; returns the label unchanged on failure."""

    if not label.startswith("xn--"):
        return label
    try:
        return label.encode("ascii").decode("idna")
    except (UnicodeError, UnicodeDecodeError):
        return label
